toolkit/tools/pkg/licensecheck/licensecheck.go

```diff
@@ -0,0 +1,411 @@
+	if err != nil {
+		return nil, fmt.Errorf("failed to walk srpm directory:\n%w", err)
+	}
+	if len(rpmsToSearchPaths) == 0 {
+		logger.Log.Warnf("No rpms found in %s", l.simpleToolChroot.ChrootRelativeMountDir())
+		return nil, nil
+	}
+
+	// Scan each rpm in parallel
+	resultsChannel := make(chan LicenseCheckResult, len(rpmsToSearchPaths))
+	cancel := make(chan struct{})
+	logger.Log.Infof("Queuing %d rpms for search", len(rpmsToSearchPaths))
+	l.queueWorkers(rpmsToSearchPaths, resultsChannel, cancel)
+	logger.Log.Infof("Searching rpms...")
+
+	// Wait for all the workers to finish, updating the progress as results come in
+	numProcessed := 0
+	lastReportPercent := 0
+	for range rpmsToSearchPaths {
+		result := <-resultsChannel
+		if result.err != nil {
+			// Signal the workers to stop if there is an error
+			err = fmt.Errorf("failed to search srpm:\n%w", result.err)
+			close(cancel)
+			return
+		}
+		numProcessed++
+		percentProcessed := (numProcessed * 100) / len(rpmsToSearchPaths)
+		if percentProcessed-lastReportPercent >= searchReportIntervalPercent {
+			logger.Log.Infof("Checked %d/%d rpms (%d%%)", numProcessed, len(rpmsToSearchPaths), percentProcessed)
+			lastReportPercent = percentProcessed
+		}
+		results = append(results, result)
+	}
+	return
+}
+
+// findRpmPaths walks the chroots's mount directory to find all *.rpm files. The paths are returned relative to the
+// chroot's root.
+func (s *LicenseChecker) findRpmPaths() (foundSrpmPaths []string, err error) {
+	const rpmExtension = ".rpm"
+	err = filepath.Walk(s.simpleToolChroot.ChrootRelativeMountDir(), func(path string, info os.FileInfo, walkErr error) error {
+		if walkErr != nil {
+			return walkErr
+		}
+		if info.IsDir() {
+			return nil
+		}
+		if !strings.HasSuffix(path, rpmExtension) {
+			return nil
+		}
+
+		foundSrpmPaths = append(foundSrpmPaths, path)
+		return nil
+	})
+	if err != nil {
+		err = fmt.Errorf("failed to walk directory:\n%w", err)
+		return nil, err
+	}
+	return foundSrpmPaths, nil
+}
+
+// queueWorkers queues up workers to search the RPMs in parallel. Each worker will wait on the jobSemaphore before starting.
+func (l *LicenseChecker) queueWorkers(rpmsToSearchPaths []string, resultsChannel chan LicenseCheckResult, cancel chan struct{}) {
+	for _, rpmPath := range rpmsToSearchPaths {
+		// Wait for the semaphore, or allow cancel before running
+		select {
+		case l.jobSemaphore <- struct{}{}:
+		case <-cancel:
+			return
+		}
+		go func(rpmPath string) {
+			defer func() {
+				<-l.jobSemaphore
+			}()
+
+			logger.Log.Debugf("Searching (%s)", filepath.Base(rpmPath))
+			searchResult, err := CheckRpmLicenses(rpmPath, l.distTag, l.licenseNames, l.exceptions)
+			logger.Log.Debugf("Finished searching (%s)", filepath.Base(rpmPath))
+			if err != nil {
+				logger.Log.Errorf("License check worker failed with error: %v", err)
+				resultsChannel <- LicenseCheckResult{err: err}
+				return
+			}
+			resultsChannel <- searchResult
```

Comment:
There's something fishy about the `LicenseCheckResult` and `CheckRpmLicenses` here. The `LicenseCheckResult` seems to combine two responsibilities:
- Holding the results of the check.
- Holding the error of the check.

This also forces us to ignore the `err` field during JSON serialization of `LicenseCheckResult`.

I'd suggest splitting the data from the error with a wrapper struct for `LicenseCheckResult` and `Error`, which can then be constructed here and sent through `resultsChannel`.

Reply:
Added a new private type
```
type licenseCheckReturn struct {
	finding LicenseCheckResult
	err     error
}
```